my-megaparsec-project/src/Statement.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Statement
+  ( Stmt(..)
+  , evalStmt
+  , prettyPrintStmt
+  , parseProgram
+  ) where
+
+import Data.Text (Text)
+import Control.Monad.State
+import qualified Data.Map as M
+import Control.Applicative ((<|>))
+import Text.Megaparsec (many, try)
+import Control.Exception (throw, Exception)
+import Data.Typeable (Typeable)
+
+import Parser
+import Expression (Expr, Subst, evalExpr, prettyPrintExpr, parseExpr)
+
+
+data IllegalArgumentException = IllegalArgumentException String
+  deriving (Show, Typeable)
+
+instance Exception IllegalArgumentException
+
+data Stmt
+  = Assign Text Expr
+  | Seq Stmt Stmt
+  | While Expr Stmt
+  | Write Expr
+  | Read Text
+  deriving (Show, Eq)
+
+evalStmt :: Stmt -> State (Subst, [Int]) [Int]
+evalStmt (Assign var e) = do
+  (subst, input) <- get
+  let e' = evalExpr e subst
+  put (M.insert var e' subst, input)
+  return []
+
+evalStmt (Seq s1 s2) = do
+  out1 <- evalStmt s1
+  out2 <- evalStmt s2
+  return (out1 ++ out2)
+
+evalStmt (Write e) = do
+  (subst, input) <- get
+  return [evalExpr e subst]
+
+evalStmt (Read var) = do
+  (subst, input) <- get
+  case input of
+    []     -> error "No more input values!"
+    (x:xs) -> do
+      put (M.insert var x subst, xs)
+      return []
+
+evalStmt (While cond body) = do
+  (subst, input) <- get
+  let condVal = evalExpr cond subst
+  if condVal < 0
+    then throw (IllegalArgumentException "While condition evaluated to negative value")
+    else if condVal == 0
+      then return []
+      else do
+        put (subst, input)
+        out1 <- evalStmt body
+        out2 <- evalStmt (While cond body)
+        return (out1 ++ out2)
+
+prettyPrintStmt :: Stmt -> String
+prettyPrintStmt (Assign var e) = show var ++ " = " ++ prettyPrintExpr e ++ ";"
+prettyPrintStmt (Seq s1 s2) = prettyPrintStmt s1 ++ "\n" ++ prettyPrintStmt s2
+prettyPrintStmt (While cond body) = "while (" ++ prettyPrintExpr cond ++ ") {\n" ++ prettyPrintStmt body ++ "\n};"
+prettyPrintStmt (Write expr) = "print(" ++ prettyPrintExpr expr ++ ");"
+prettyPrintStmt (Read var) = show var ++ " = readInt();"
```

Comment:
A pretty-printing library such as [prettyprinter](https://hackage.haskell.org/package/prettyprinter) can help to make your code  ✨ pretty ✨